File: leetcode/lc-us/cpp/865.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode *subtreeWithAllDeepest(TreeNode *root) {
        return dfs(root).first;
    }

private:
    pair<TreeNode *, int> dfs(TreeNode *node) {
        if (!node) return {nullptr, 0};
        auto l = dfs(node->left), r = dfs(node->right);
        if (l.second == r.second) return {node, l.second + 1};
        else if (l.second > r.second) return {l.first, l.second + 1};
        else return {r.first, r.second + 1};
    }
};

File: leetcode/lc-us/py/865.py
import collections


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        Node = collections.namedtuple("Node", ("node", "max_depth"))

        def dfs(node: TreeNode) -> Node:
            if not node:
                return Node(None, 0)
            left = dfs(node.left)
            right = dfs(node.right)
            if left[1] == right[1]:
                return Node(node, left[1] + 1)
            elif left[1] > right[1]:
                return Node(left[0], left[1] + 1)
            else:
                return Node(right[0], right[1] + 1)

        return dfs(root).node
